Plugins/BuildServerIntegration/JenkinsIntegration/JenkinsAdapter.cs

```diff
@@ -81,67 +82,195 @@ public void Initialize(IBuildServerWatcher buildServerWatcher, ISettingsSource c
+                            {
+                                //Freestyle jobs
+                                s = jobDescription["builds"];
+                            }
+                            else if (jobDescription["jobs"] != null)
+                            {
+                                //Multibranch pipeline
+                                s = jobDescription["jobs"]
+                                    .SelectMany(j => j["builds"]);
+                                foreach (var j in jobDescription["jobs"])
+                                {
+                                    long ts = j["lastBuild"]["timestamp"].ToObject<long>();
+                                    timestamp = Math.Max(timestamp, ts);
+                                }
+                            }
+                            //else: The server had no response (overloaded?) or a multibranch pipeline is not configured
+
+                            if (jobDescription["lastBuild"] != null)
+                            {
+                                timestamp = jobDescription["lastBuild"]["timestamp"].ToObject<long>();
+                            }
+                        }
+
+                        return new ResponseInfo
+                        {
+                            Url = projectUrl,
+                            Timestamp = timestamp,
+                            JobDescription = s
+                        };
+                    },
+                    TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.AttachedToParent);
         }
 
         public IObservable<BuildInfo> GetFinishedBuildsSince(IScheduler scheduler, DateTime? sinceDate = null)
         {
-            return GetBuilds(scheduler, sinceDate, false);
+            //GetBuilds() will return the same builds as for GetRunningBuilds().
+            //Multiple calls will fetch same info multiple times and make debugging very confusing
+            //Similar as for AppVeyor
+            return Observable.Empty<BuildInfo>();
         }
 
         public IObservable<BuildInfo> GetRunningBuilds(IScheduler scheduler)
         {
             return GetBuilds(scheduler, null, true);
         }
 
-        public IObservable<BuildInfo> GetBuilds(IScheduler scheduler, DateTime? sinceDate = null, bool? running = null)
+        private IObservable<BuildInfo> GetBuilds(IScheduler scheduler, DateTime? sinceDate = null, bool? running = null)
         {
             return Observable.Create<BuildInfo>((observer, cancellationToken) =>
                 Task<IDisposable>.Factory.StartNew(
                     () => scheduler.Schedule(() => ObserveBuilds(sinceDate, running, observer, cancellationToken))));
         }
 
+        //Jenkins builds are retrieved with latest build first
+        //If there are many builds on the same commit, it is normally the latest that is interesting, ignore the following
+        private bool DifferFromPrev(BuildInfo curr, ref BuildInfo prev)
```

Comment:
You should use HashSet here, as we can not assume that rebuilds for a given commit are consecutive.

Reply:
Can change, this was intended as an improvement for the common case rather than fool proof. Hash will also consistently select first matching project.